File: skills/code-architect/references/code-examples/go/bridge.go
// Bridge: decouple an abstraction from its implementation so the two can
// vary independently.
package main

import "fmt"

// Implementor is the low-level interface the abstraction delegates to.
type Implementor interface {
	Emit(message string) string
}

type plainBackend struct{}

func (plainBackend) Emit(m string) string { return m }

type taggedBackend struct{}

func (taggedBackend) Emit(m string) string { return "[" + m + "]" }

// Abstraction holds a reference to an Implementor (the bridge).
type Abstraction struct {
	impl Implementor
}

func (a Abstraction) Show(text string) string {
	return a.impl.Emit(text)
}

// RefinedAbstraction extends behaviour without touching implementors.
type RefinedAbstraction struct {
	Abstraction
}

func (r RefinedAbstraction) ShowTwice(text string) string {
	return r.impl.Emit(text) + " " + r.impl.Emit(text)
}

func main() {
	a := Abstraction{impl: plainBackend{}}
	fmt.Println(a.Show("ping"))

	r := RefinedAbstraction{Abstraction{impl: taggedBackend{}}}
	fmt.Println(r.ShowTwice("ping"))
}

File: skills/code-architect/references/code-examples/rust/bridge.rs
// Bridge — decouple an abstraction from its implementation so the two can vary
// independently; the abstraction holds a reference to an implementor.

// Implementor hierarchy.
trait Renderer {
    fn draw_label(&self, text: &str) -> String;
}

struct PlainRenderer;
struct FancyRenderer;

impl Renderer for PlainRenderer {
    fn draw_label(&self, text: &str) -> String {
        text.to_string()
    }
}
impl Renderer for FancyRenderer {
    fn draw_label(&self, text: &str) -> String {
        format!("*** {} ***", text)
    }
}

// Abstraction hierarchy holds a renderer (the bridge).
struct Button {
    renderer: Box<dyn Renderer>,
    caption: String,
}
struct Badge {
    renderer: Box<dyn Renderer>,
    count: u32,
}

impl Button {
    fn show(&self) -> String {
        format!("[{}]", self.renderer.draw_label(&self.caption))
    }
}
impl Badge {
    fn show(&self) -> String {
        self.renderer.draw_label(&format!("({})", self.count))
    }
}

fn main() {
    let plain_btn = Button { renderer: Box::new(PlainRenderer), caption: "Save".into() };
    let fancy_btn = Button { renderer: Box::new(FancyRenderer), caption: "Save".into() };
    let fancy_badge = Badge { renderer: Box::new(FancyRenderer), count: 7 };

    println!("{}", plain_btn.show());
    println!("{}", fancy_btn.show());
    println!("{}", fancy_badge.show());
}
